src/instructlab/configuration.py

```diff
@@ -619,7 +644,228 @@ def ensure_storage_directories_exist():
         "lora_target_modules": ["q_proj", "k_proj", "v_proj", "o_proj"],
     }
 
-    # create exper_args file for users to see/edit
+    TRAIN_DIR_EXPECTED_FILES = {
+        "A100_H100_x4.yaml",
+        "A100_H100_x8.yaml",
+        "A100_H100_x2.yaml",
+        "L40_x8.yaml",
+        "L40_x4.yaml",
+        "L4_x8.yaml",
+    }
+
+    profile_dir = os.environ.get(DEFAULTS.ILAB_TRAIN_PROFILE_DIR)
+    if profile_dir != "" and profile_dir is not None:
+        # the train dir exists, read it
+        # expect only the supported cfgs.
+        for file in TRAIN_DIR_EXPECTED_FILES:
+            new_file = os.path.join(DEFAULTS.TRAIN_PROFILE_DIR, file)
+            tmpl_file = os.path.join(profile_dir, file)
+            train_cfg = read_train_profile(tmpl_file)
+            if not os.path.isfile(new_file):
+                with open(new_file, "w", encoding="utf-8") as outfile:
+                    d = train_cfg.model_dump_json()
+                    loaded = yaml.load(d, Loader=yaml.SafeLoader)
+                    yaml.dump(loaded, outfile)
+    else:
+        # else render the defaults
+        FOUR_GPU_TRAIN_AH = _train(
+            additional_args={
+                "warmup_steps": 10,
+                "learning_rate": 2e-5,
+                "lora_dropout": 0.1,
+                "lora_alpha": 32,
+            },
+            ckpt_output_dir=os.path.join(DEFAULTS._data_dir, "checkpoints"),
+            data_output_dir=os.path.join(DEFAULTS._data_dir, "internal"),
+            data_path=os.path.join(DEFAULTS._data_dir, "datasets"),
+            nproc_per_node=4,
+            effective_batch_size=96,
+            lora_quantize_dtype=None,
+            lora_rank=0,
+            max_batch_len=60000,
+            max_seq_len=4096,
+            save_samples=1000,
+            deepspeed_cpu_offload_optimizer=False,
+            is_padding_free=True,
+            num_epochs=7,
+            model_path=os.path.join(
+                DEFAULTS._cache_home, "models/instructlab/granite-7b-lab"
+            ),
+        )
+
+        EIGHT_GPU_TRAIN_AH = _train(
```

Comment:
These configs have a lot of shared values. Would it define a single base config and additional_params and override each of these config objects with the values that change between them? It'd be easier to identify what the major changes are between the profiles and make refactoring a lot easier.

Reply:
This was my initial thought, but for the sake of getting this across I think we should handle this in a follow-up PR. It will involve additional folding logic that we'd need to bring in, which expands the scope of this PR.

Reply:
I'm not sure how this'd work given that not everything is in `additional_args`. If we were to define a base _train class and then create other _train classes, I am not aware of a way to unwrap one into the other? I might be missing something though

Reply:
You could probably have some dicts with common and specific settings, and then

```
base_config = { ... }
config1 = { ... }
config2 = { ... }
final_config = _train(**(base_config | config1))
```

Reply:
Pydantic objects aren't immutable by default.

```
base_config = _train(**defaults)
EIGHT_L_FOUR_GPU = base_config.model_copy(deep=True)
EIGHT_L_FOUR_GPU.nproc_per_node = 8
EIGHT_L_FOUR_GPU.max_batch_len = 30000
EIGHT_L_FOUR_GPU.lora_rank = 0
EIGHT_L_FOUR_GPU.deepspeed_cpu_offload_optimizer=True
```

There's a tradeoff with this, which is that the input to the objects aren't as nicely formed, but you could do something like:

```
base_config = _train(**defaults)
EIGHT_GPU_TRAIN_AH = base_config.model_copy(deep=True)
EIGHT_GPU_TRAIN_AH.nproc_per_node = 8
# ... any overrides for eight GPUs of A or H

FOUR_GPU_TRAIN_AH = EIGHT_GPU_TRAIN_AH.model_copy(deep=True)
FOUR_GPU_TRAIN_AH.nproc_per_node = 4
```